src/ut/baseresolver_test.cpp

```diff
@@ -305,6 +316,51 @@ TEST_F(BaseResolverTest, ARecordMakeUpUsedGray)
   EXPECT_THAT(targets, Contains(gray_record));
 }
 
+/// Test that the lazy target selection iterator returns true when the target
+/// has been set, and false otherwise
+TEST_F(BaseResolverTest, ARecordIteratorNextReturnValue)
+{
+  AddrInfo record;
+  AddrInfo expected_record = ResolverTest::ip_to_addr_info("3.0.0.0");
+
+  add_white_records(1);
+  BaseAddrIterator* it = resolve_iter();
+
+  // The value of record should be set by the iterator.
+  EXPECT_TRUE(it->next(record));
+  EXPECT_EQ(record, expected_record);
+
+  // The value of record should be left unchanged by the iterator.
+  EXPECT_FALSE(it->next(record));
+  EXPECT_EQ(record, expected_record);
+
+  delete it; it = nullptr;
+}
+
+/// Test that the lazy target selection iterator uses the state of each Host at
+/// the time that next is called each time, i.e. acts lazily
+TEST_F(BaseResolverTest, ARecordIteratorIsLazy)
+{
+  add_white_records(3);
+
+  // Blacklist a record
+  AddrInfo black_to_gray_record = ResolverTest::ip_to_addr_info("3.0.0.0");
+  _baseresolver.blacklist(black_to_gray_record);
+
+  // Get an iterator
+  BaseAddrIterator* it = resolve_iter();
+
+  // Move the record to the graylist
+  cwtest_advance_time_ms(31000);
+
+  // The record should be returned by the next call to the iterator
+  AddrInfo record;
+  EXPECT_TRUE(it->next(record));
+  EXPECT_EQ(record, black_to_gray_record);
+
+  delete it; it = nullptr;
+}
```

Comment:
You could make this test more convincing by creating two iterators and consuming one before you advance time and the other afterwards. I know the first iterator could return an acceptable address by chance, but this testcase would make it really clear what the point of lazy iteration is.